File: backend/app/api/v1/ws.py
"""Authenticated WebSocket endpoint.

JWT is passed in the query string (?token=...) because browser WebSocket
clients can't easily set Authorization headers.

Protocol (client -> server JSON messages):
  {"action": "subscribe",   "symbols": ["AAPL", "MSFT"]}
  {"action": "unsubscribe", "symbols": ["AAPL"]}
  {"action": "list"}
  {"action": "ping"}

Server -> client:
  {"type": "connected",  "user_id": 1, "subscriptions": ["AAPL", ...]}
  {"type": "subscribed", "subscriptions": [...]}
  {"type": "unsubscribed", "subscriptions": [...]}
  {"type": "list",       "subscriptions": [...]}
  {"type": "pong"}
  {"type": "price",      "data": {symbol, price, change, change_percent, ...}}
  {"type": "alert",      "data": {rule_id, symbol, message, ...}}
  {"type": "error",      "message": "..."}
"""
from __future__ import annotations

import logging

import jwt
from fastapi import APIRouter, Query, WebSocket, WebSocketDisconnect, status

from app.core.security import decode_token
from app.db.session import AsyncSessionLocal
from app.services import watchlist_service
from app.services.user_service import get_user_by_id
from app.streaming.ws_hub import get_ws_hub

log = logging.getLogger(__name__)
router = APIRouter()


@router.websocket("/ws")
async def websocket_endpoint(
    websocket: WebSocket,
    token: str = Query(..., description="JWT access token"),
) -> None:
    # ---- Auth (before accept, so unauthorized clients get a clean close) ----
    try:
        payload = decode_token(token)
        if payload.get("type") != "access":
            raise ValueError("Not an access token")
        user_id = int(payload["sub"])
    except (jwt.PyJWTError, KeyError, TypeError, ValueError) as e:
        log.warning("WS auth failed: %s", e)
        await websocket.close(code=status.WS_1008_POLICY_VIOLATION)
        return

    async with AsyncSessionLocal() as db:
        user = await get_user_by_id(db, user_id)
        if user is None or not user.is_active:
            await websocket.close(code=status.WS_1008_POLICY_VIOLATION)
            return
        # Auto-subscribe to current watchlist
        items = await watchlist_service.list_for_user(db, user)
        initial_symbols = [it.symbol for it in items]

    await websocket.accept()
    hub = get_ws_hub()
    await hub.connect(websocket, user_id)
    if initial_symbols:
        await hub.subscribe(user_id, initial_symbols)

    try:
        await websocket.send_json(
            {
                "type": "connected",
                "user_id": user_id,
                "subscriptions": sorted(await hub.list_subscriptions(user_id)),
            }
        )

        while True:
            try:
                msg = await websocket.receive_json()
            except WebSocketDisconnect:
                break
            except Exception as e:
                await websocket.send_json(
                    {"type": "error", "message": f"Invalid message: {e}"}
                )
                continue

            action = (msg.get("action") or "").lower()
            if action == "subscribe":
                symbols = msg.get("symbols", []) or []
                subs = await hub.subscribe(user_id, symbols)
                await websocket.send_json(
                    {"type": "subscribed", "subscriptions": sorted(subs)}
                )
            elif action == "unsubscribe":
                symbols = msg.get("symbols", []) or []
                subs = await hub.unsubscribe(user_id, symbols)
                await websocket.send_json(
                    {"type": "unsubscribed", "subscriptions": sorted(subs)}
                )
            elif action == "list":
                subs = await hub.list_subscriptions(user_id)
                await websocket.send_json(
                    {"type": "list", "subscriptions": sorted(subs)}
                )
            elif action == "ping":
                await websocket.send_json({"type": "pong"})
            else:
                await websocket.send_json(
                    {"type": "error", "message": f"Unknown action: {action}"}
                )
    finally:
        await hub.disconnect(websocket, user_id)
        log.debug("WS disconnected: user=%s", user_id)

File: backend-ts/src/routes/ws.ts
/**
 * WebSocket hub endpoint.
 *
 *   GET ws://host/api/v1/ws?token=<access_jwt>
 *
 * Auto-subscribes the connected user to every symbol on their watchlist.
 * Client can subscribe/unsubscribe to additional symbols at any time:
 *   { action: "subscribe",   symbols: ["MSFT","TSLA"] }
 *   { action: "unsubscribe", symbols: ["TSLA"] }
 */
import type { FastifyInstance } from "fastify";
import { prisma } from "../db.js";
import { decodeToken } from "../lib/security.js";
import {
  register,
  subscribe,
  unsubscribe,
} from "../services/hub.js";

export default async function wsRoutes(app: FastifyInstance) {
  app.get("/ws", { websocket: true }, async (socket, req) => {
    const token = (req.query as { token?: string })?.token;
    if (!token) {
      socket.send(JSON.stringify({ type: "error", detail: "missing token" }));
      socket.close();
      return;
    }
    let userId: number;
    try {
      const claims = decodeToken(token);
      if (claims.type !== "access") throw new Error("not access token");
      userId = Number.parseInt(claims.sub, 10);
    } catch {
      socket.send(JSON.stringify({ type: "error", detail: "invalid token" }));
      socket.close();
      return;
    }

    // Auto-subscribe to the user's watchlist on connect.
    const items = await prisma.watchlistItem.findMany({ where: { userId } });
    const initial = items.map((it) => it.symbol);
    const client = register(socket as never, userId, initial);

    socket.send(
      JSON.stringify({ type: "connected", user_id: userId, subscriptions: initial }),
    );

    socket.on("message", (raw) => {
      try {
        const msg = JSON.parse(raw.toString());
        if (msg.action === "subscribe" && Array.isArray(msg.symbols)) {
          subscribe(client, msg.symbols);
          socket.send(
            JSON.stringify({ type: "subscribed", symbols: Array.from(client.subscriptions) }),
          );
        } else if (msg.action === "unsubscribe" && Array.isArray(msg.symbols)) {
          unsubscribe(client, msg.symbols);
          socket.send(
            JSON.stringify({ type: "unsubscribed", symbols: Array.from(client.subscriptions) }),
          );
        } else if (msg.action === "ping") {
          socket.send(JSON.stringify({ type: "pong" }));
        }
      } catch {
        /* ignore malformed client frame */
      }
    });
  });
}
